Repository methods with type safe inclusion

HlibBondarev opened the pull request:
1) Changed the signature of IncludeProperties() extension method in QueryableExtensions class -
added parameter includeExpression with type - Func<IQueryable<T>, IQueryable<T>>;
2) Changed the signature of methods in IEntityAddOnlyRepository interface and classes (EntityRepositoryBase, EntityRepositorySoftDeleted) implemented this interface to use the changed IncludeProperties() extension method;
3) Changed all services that used methods implemented IEntityAddOnlyRepository interface;
4) Fixed appropriated tests.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
  - Enhanced data retrieval logic to improve filtering consistency and ensure robust inclusion of related information across features like workshops, applications, and user records.
  - Introduced new parameters for flexible query manipulation in multiple repository methods, enhancing the clarity and maintainability of data access.
- **Tests**
  - Updated testing configurations to align with the new query patterns, boosting reliability and validating overall system stability.
  - Added new test methods to verify functionality and ensure comprehensive coverage of recent changes.
- **Chore**
  - Performed various code cleanups and formatting improvements for a more streamlined and maintainable codebase.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HlibBondarev committed:
1) Changed the signature of IncludeProperties() extension method in QueryableExtensions class -
added parameter includeExpression with type - Func<IQueryable<T>, IQueryable<T>>.
2) Changed the signature of methods in IEntityAddOnlyRepository interface and classes (EntityRepositoryBase, EntityRepositorySoftDeleted) implemented this interface to use the changed IncludeProperties() extension method.
3) Changed all services that used methods implemented IEntityAddOnlyRepository interface.
4) Fixed appropriated tests.

HlibBondarev commented:
@coderabbitai review

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ChangesLogService.cs:
```diff
@@ -302,7 +302,7 @@ public async Task<SearchResult<EmployeeChangesLogDto>> GetEmployeeChangesLogAsyn
         var count = await employeeChangesLogRepository.Count(where).ConfigureAwait(false);
         var query = employeeChangesLogRepository
-            .Get(request.From, request.Size, string.Empty, where, sortExpression, true)
+            .Get(request.From, request.Size, string.Empty, null, where, sortExpression, true)
```
this comment is for every similar method usage - it it better o add argument names instead of passing null. Tha way we can even remove `string.Empty` in some cases. Tests (moq setup) is the only exception where you will have to pass `null` explicitly.

HlibBondarev replied:
Ok, corrected

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -213,7 +220,8 @@ public async Task<SearchResult<WorkshopDto>> GetAll(OffsetFilter offsetFilter)
                     skip: offsetFilter.From,
                     take: offsetFilter.Size,
-                    includeProperties: includingPropertiesForMappingDtoModel,
+                    //includeProperties: includingPropertiesForMappingDtoModel,
```
if you switch to new model - don't leave commented code

HlibBondarev replied:
ok

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -301,6 +309,7 @@ public async Task<SearchResult<WorkshopProviderViewCard>> GetByProviderId(Guid i
                 skip: filter.From,
                 take: filter.Size,
                 includeProperties: includingPropertiesForMappingDtoModel,
```
probably should remove string?

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Extensions/QueryableExtensions.cs:
```diff
@@ -9,10 +9,18 @@ public static class QueryableExtensions
+
+        if (includeExpression != null)
+        {
+            _ = includeExpression?.Invoke(query);
```
shouldn't it be `query = includeExpression?.Invoke(query)`. Also you can use it simply like `includeExpression(query)`

HlibBondarev replied:
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/PositionServiceTests.cs:
```diff
@@ -331,8 +342,11 @@ public async Task DeletePosition_WhenIsDeletedTrue_ReturnsExceptionMessage()
 
-        _mockRepository.Setup(r => r.GetByFilter(It.IsAny<Expression<Func<Position, bool>>>(), It.IsAny<string>()))
-        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties) =>
+        _mockRepository.Setup(r => r.GetByFilter(
+            It.IsAny<Expression<Func<Position, bool>>>(), 
+            It.IsAny<string>(),
+            It.IsAny<Func<IQueryable<Position>, IQueryable<Position>>>()))
+        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties, Func<IQueryable<Parent>, IQueryable<Parent>> includeExpression) =>
```
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/PositionServiceTests.cs:
```diff
@@ -303,8 +311,11 @@
         
-        _mockRepository.Setup(r => r.GetByFilter(It.IsAny<Expression<Func<Position, bool>>>(), It.IsAny<string>()))
-        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties) =>
+        _mockRepository.Setup(r => r.GetByFilter(
+            It.IsAny<Expression<Func<Position, bool>>>(),
+            It.IsAny<string>(),
+            It.IsAny<Func<IQueryable<Position>, IQueryable<Position>>>()))
+        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties, Func<IQueryable<Parent>, IQueryable<Parent>> includeExpression) =>
```
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/PositionServiceTests.cs:
```diff
@@ -192,8 +197,11 @@
                         
-        _mockRepository.Setup(r => r.GetByFilter(It.IsAny<Expression<Func<Position, bool>>>(), It.IsAny<string>()))
-        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties) =>
+        _mockRepository.Setup(r => r.GetByFilter(
+            It.IsAny<Expression<Func<Position, bool>>>(),
+            It.IsAny<string>(),
+            It.IsAny<Func<IQueryable<Position>, IQueryable<Position>>>()))
+        .ReturnsAsync((Expression<Func<Position, bool>> predicate, string includeProperties, Func<IQueryable<Parent>, IQueryable<Parent>> includeExpression) =>
```
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.DataAccess/Extensions/QueryableExtensions.cs:
```diff
@@ -9,10 +9,18 @@ public static class QueryableExtensions
     {
         ArgumentNullException.ThrowIfNull(query);
+
+        if (includeExpression != null)
+        {
+            _ = includeExpression?.Invoke(query);
+
+            return query;
+        }
```
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.DataAccess/Repository/Base/Api/IEntityAddOnlyRepository.cs:
```diff
@@ -69,35 +69,50 @@
+    IQueryable<TEntity> GetByFilterNoTracking(
+        Expression<Func<TEntity, bool>> whereExpression, 
+        string includeProperties = "",
+        Func<IQueryable<TEntity>, IQueryable<TEntity>> includeExpression = null);
```
Ok

HlibBondarev committed:
Fixed the bugs mentioned Dmytro Minochkin and CodeRabbit.

HlibBondarev committed:
1) changed includeFunc private property in WorkshopService class;
2) changed GetByFilterNoTracking() method of EntityRepositoryBase class - added includeExpression parameter;
3) changed Delete_SoftDeletes_Provider_DeletesRelatedEntities method in ProviderRepositoryTests class - added includeExpression parameter.

HlibBondarev committed:
1) Added named parameters when using ChildRepository.Get() method in methods of ChildService class;
2) Added class ChildRepositoryTests.

HlibBondarev committed:
1) Fixed GetByFilter() method of ChildService class.
2) Added test to GetByFilter() method of ChildService class.

HlibBondarev committed:
Create tests for methods (GetByParentIdOrderedByFirstName and GetByUserId) of ChildService class.

HlibBondarev committed:
Fixed test Delete_SoftDeletes_Child_DeletesRelatedEntities in ChildRepositoryTests class.
